### PR TITLE
removed lifetimes 

### PR DESCRIPTION
**Pull request by debris**
had difficulties using `send_transaction_with_confirmation` so I removed lifetimes from `EthFilter::stream`, `wait_for_confirmations` and `send_transaction_with_confirmation`

### CODE REVIEW

File: src/api/mod.rs [hunk 1]
```diff
@@ -80,19 +80,17 @@ impl<T: Transport> Web3Main<T> {
   }
 
   /// Should be used to wait for confirmations
-  pub fn wait_for_confirmations<'a, F, V>(&'a self, poll_interval: Duration, confirmations: usize, check: V)
-  -> Box<Future<Item = (), Error = Error> + 'a> where
-  F: 'a + IntoFuture<Item = Option<U256>, Error = Error>,
-  V: 'a + Fn() -> F,
+  pub fn wait_for_confirmations<F, V>(&self, poll_interval: Duration, confirmations: u64, check: V)
+    -> confirm::Confirmations<&T, V, F::Future> where
```
Comment (tomusdrw) [inline src/api/mod.rs]: Style nitpick: I would rather see the output type on the same line as the function name and type parameters idented.
If it doesn't fit then it should be split by parameters:
```rust
pub fn wait_for_confirmations<F, V>(
  &self,
  poll_interval: Duration,
  confirmations: u64,
  check: V,
) -> confirm::Confirmations<&T, V, F::Future> where
  F: IntoFuture<..>,
  V: confirm::ConfirmationCheck<Check = F>,
{
 ...
}
```

File: src/confirm.rs [hunk 2]
```diff
@@ -1,48 +1,230 @@
...
-  let eth = Eth::new(transport.clone());
-  let result = eth.send_transaction(tx)
-    .and_then(move |hash| {
-      wait_for_confirmations(transport.clone(), poll_interval, confirmations, move || {
-        eth.transaction_receipt(hash.clone()).map(|option| option.map(|receipt| receipt.block_number))
-      })
-      .and_then(move |_| {
-        let eth = Eth::new(transport.clone());
-        eth.transaction_receipt(hash).and_then(|option| option.ok_or(Error::Unreachable))
+pub trait ConfirmationCheck {
```
Comment (tomusdrw) [inline src/confirm.rs]: Should it have a blanket implementation for `T, F where F: Fn() -> T, T: IntoFuture<Item=Option<U256>, Error>`?

File: src/confirm.rs [hunk 3]
```diff
@@ -1,48 +1,230 @@
...
+  state: WaitForConfirmationsState<F, T::Out>,
+  filter_stream: Skip<FilterStream<T, H256>>,
+  confirmation_check: V,
+  confirmations: u64,
+}
+
+impl<T, V, F> Future for WaitForConfirmations<T, V, F::Future> where
+  T: Transport,
+  V: ConfirmationCheck<Check = F>,
+  F: IntoFuture<Item = Option<U256>, Error = Error> {
```
Comment (tomusdrw) [inline src/confirm.rs]: Style nitpick: missing trailing `,` and `{` should be on a new line.

File: src/confirm.rs [hunk 4]
```diff
@@ -1,48 +1,230 @@
...
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    loop {
+      let next_state = match self.state {
+        WaitForConfirmationsState::WaitForNextBlock => {
+          let _ = try_ready!(self.filter_stream.poll());
+          WaitForConfirmationsState::CheckConfirmation(self.confirmation_check.check().into_future())
+        },
+        WaitForConfirmationsState::CheckConfirmation(ref mut future) => match try_ready!(future.poll()) {
+          Some(confirmation_block_number) => {
+            let future = Eth::new(&self.transport).block_number();
```
Comment (tomusdrw) [inline src/confirm.rs]: Can't we do it in parallel with `confirmation_check.check()`?

File: src/confirm.rs [hunk 5]
```diff
@@ -1,48 +1,230 @@
...
+  Wait(WaitForConfirmations<T, V, F>),
+}
+
+pub struct Confirmations<T: Transport, V, F> {
+  state: ConfirmationsState<T, V, F>,
+}
+
+impl<T: Transport + Clone, V, F> Confirmations<T, V, F> {
+  fn new(transport: T, poll_interval: Duration, confirmations: u64, check: V) -> Self {
+    let eth = EthFilter::new(transport.clone());
```
Comment (tomusdrw) [inline src/confirm.rs]: I think this could be just `&transport` here (and `+Clone` bound would be redundant)

File: src/confirm.rs [hunk 6]
```diff
@@ -1,48 +1,230 @@
...
-    });
-  Box::new(result)
+    }
+  }
+}
+
+impl<T, V, F> Future for Confirmations<T, V, F::Future> where
+  T: Transport,
+  V: ConfirmationCheck<Check = F>,
+  F: IntoFuture<Item = Option<U256>, Error = Error> {
```
Comment (tomusdrw) [inline src/confirm.rs]: style nitpick (comma and separate line)

File: src/confirm.rs [hunk 7]
```diff
@@ -1,48 +1,230 @@
...
+      };
+      self.state = next_state;
+    }
+  }
+}
+
+struct CreateWaitForConfirmations<T: Transport, V> {
+  create_filter: CreateFilter<T, H256>,
+  poll_interval: Duration,
+  transport: Option<T>,
```
Comment (tomusdrw) [inline src/confirm.rs]: We `.take()` both of them, so maybe better store it as `Option<(T, V)>` to have only one `expect()`?

File: src/confirm.rs [hunk 8]
```diff
@@ -1,48 +1,230 @@
...
+      };
+      self.state = next_state;
+    }
+  }
+}
+
+pub fn wait_for_confirmations<T, V, F>(transport: T, poll_interval: Duration, confirmations: u64, check: V) -> Confirmations<T, V, F::Future> where
+  T: Transport + Clone,
+  V: ConfirmationCheck<Check = F>,
+  F: IntoFuture<Item = Option<U256>, Error = Error>, {
```
Comment (tomusdrw) [inline src/confirm.rs]: style nitpick - `{` in a separate line

File: src/confirm.rs [hunk 9]
```diff
@@ -1,48 +1,230 @@
...
+  state: SendTransactionWithConfirmationState<T>,
+  eth: Eth<T>,
+  transport: T,
+  poll_interval: Duration,
+  confirmations: u64,
+}
+
+impl<T: Transport + Clone> SendTransactionWithConfirmation<T> {
+  fn new(transport: T, tx: TransactionRequest, poll_interval: Duration, confirmations: u64) -> Self {
+  let eth = Eth::new(transport.clone());
```
Comment (tomusdrw) [inline src/confirm.rs]: Bad indentation.

File: src/confirm.rs [hunk 10]
```diff
@@ -1,48 +1,230 @@
...
+  type Item = TransactionReceipt;
+  type Error = Error;
+
+  fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+    loop {
+      let next_state = match self.state {
+        SendTransactionWithConfirmationState::SendTransaction(ref mut future) => {
+          let hash = try_ready!(future.poll());
+          let confirmation_check = TransactionReceiptBlockNumberCheck::new(Eth::new(self.transport.clone()), hash.clone());
+          let wait = wait_for_confirmations(self.transport.clone(), self.poll_interval, self.confirmations, confirmation_check);
```
Comment (tomusdrw) [inline src/confirm.rs]: Isn't it enough to pass `&self.transport`?

File: src/lib.rs [hunk 11]
```diff
@@ -75,7 +75,7 @@ pub type RequestId = usize;
 /// Transport implementation
 pub trait Transport {
   /// The type of future this transport returns when a call is made.
-  type Out: futures::Future<Item=rpc::Value, Error=Error>;
+  type Out: futures::Future<Item=rpc::Value, Error=Error> + 'static;
```
Comment (tomusdrw) [inline src/lib.rs]: I think that's not really needed (remember we were just trying this out).

File: src/confirm.rs (hunk 2 above)
Comment (debris) [inline src/confirm.rs]: imo, there is no need to right now

File: src/confirm.rs (hunk 5 above)
Comment (debris) [inline src/confirm.rs]: I can't do that without using lifetimes, cause then `Confirmations` would contain 2 transports: `T` && `&T`.

File: src/confirm.rs (hunk 10 above)
Comment (debris) [inline src/confirm.rs]: same as above

Comment (debris) [inline src/confirm.rs]: same as above

File: src/confirm.rs (hunk 4 above)
Comment (debris) [inline src/confirm.rs]: we can, but then we make redundant calls to get block number when `confirmation_check.check()` returns `None`

File: src/confirm.rs (hunk 7 above)
Comment (debris) [inline src/confirm.rs]: I agree, but in the same time, putting them in a tuple will be less descriptive

File: src/lib.rs [hunk 12]
```diff
@@ -31,7 +31,7 @@ pub mod contract;
 pub mod transports;
 pub mod types;
 
-mod confirm;
+pub mod confirm;
```
Comment (tomusdrw) [inline src/lib.rs]: If it's public it should be aligned with other modules.

File: src/confirm.rs [hunk 13]
```diff
@@ -1,48 +1,244 @@
...
+use types::{H256, U256, TransactionRequest, TransactionReceipt};
+use helpers::CallResult;
 use {Transport, Error};
 
-pub fn wait_for_confirmations<'a, T, F, V>(transport: T, poll_interval: Duration, confirmations: usize, check: V)
-  -> Box<Future<Item = (), Error = Error> + 'a> where
-  T: 'a + Transport + Clone,
-  F: 'a + IntoFuture<Item = Option<U256>, Error = Error>,
-  V: 'a + Fn() -> F,
+pub trait ConfirmationCheck {
```
Comment (tomusdrw) [inline src/confirm.rs]: Since the module is now public all public items require a documentation.

File: src/types/transaction.rs [hunk 14]
```diff
@@ -34,10 +34,6 @@ pub struct Transaction {
 /// "Receipt" of an executed transaction: details of its execution.
 #[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize)]
 pub struct Receipt {
-  /// Hash
-  pub hash: H256,
```
Comment (tomusdrw) [inline src/types/transaction.rs]: Should be in a separate PR

**Comment by debris** [comment]
of course :)
